[WPE] Add support for generating a cross toolchain, images and cross building easily for a set of curated embedded targets with Yocto

clopez opened the pull request:
#### 80926377d6a9f2454a3fd3c5f586b2a8f6c7ad34
<pre>
[WPE] Add support for generating a cross toolchain, images and cross building easily for a set of curated embedded targets with Yocto
<a href="https://bugs.webkit.org/show_bug.cgi?id=249604">https://bugs.webkit.org/show_bug.cgi?id=249604</a>

Reviewed by Philippe Normand.

This adds a new script named &apos;cross-toolchain-helper&apos; that
automates the process of building a cross-toolchain and an
image with Yocto.

The script also is able to automatically deploy the Yocto
workdir and set the target build configs from a set of
configuration files.

The main configuration file for this script is at the path
&apos;Tools/yocto/targets.conf&apos;. Inside this file several targets
can be defined. Each target defines a set of Yocto layers
and local.conf configurations. It can be also defined the
default WebKit parameters for each target.

The following WebKit tool scripts are also patched to be
aware of the cross-target builds:
  - Tools/Scripts/update-webkitwpe-libs
  - Tools/Scripts/update-webkitgtk-libs
  - Tools/Scripts/build-webkit
  - Tools/CISupport/built-product-archive

To enable the cross-target builds the flag &apos;--cross-target=target_name&apos;
should be passed. Alternatively the environment variable
&apos;WEBKIT_CROSS_TARGET&apos; can be used instead.

The list of targets defined at &apos;Tools/yocto/targets.conf&apos;
is curated and is expected to build and work without issues.
In the near future a post-commit bot will be checking this.

The tool is designed to automatically wipe and recreate the
cross-toolchain and images when it detects the configuration
for any target has changed. However, this behaviour can be
overriden by optional parameters.

For more info (including examples on how to use this) check the file:
Tools/yocto/README.md

* Tools/CISupport/built-product-archive:
* Tools/Scripts/build-webkit:
* Tools/Scripts/cross-toolchain-helper: Added.
* Tools/Scripts/update-webkitgtk-libs:
* Tools/Scripts/update-webkitwpe-libs:
* Tools/Scripts/webkitdirs.pm:
(determineCrossTarget):
(determineConfigurationProductDir):
(determinePassedConfiguration):
(runInCrossTargetEnvironment):
(inCrossTargetEnvironment):
(getCrossTargetName):
(shouldBuildForCrossTarget):
(wrapperPrefixIfNeeded):
(shouldUseFlatpak):
(cmakeCachePath):
(cmakeFilesPath):
(shouldRemoveCMakeCache):
(cmakeGeneratedBuildfile):
(generateBuildSystemFromCMakeProject):
(productDirForCMake):
(buildCMakeGeneratedProject):
(cleanCMakeGeneratedProject):
* Tools/yocto/README.md: Added.
* Tools/yocto/rpi/bblayers.conf: Added.
* Tools/yocto/rpi/fix-nativesdk-image-defs.patch: Added.
* Tools/yocto/rpi/local-rpi3-32bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi3-32bits-userland.conf: Added.
* Tools/yocto/rpi/local-rpi3-64bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi4-32bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi4-64bits-mesa.conf: Added.
* Tools/yocto/rpi/manifest.xml: Added.
* Tools/yocto/targets.conf: Added.

Canonical link: <a href="https://commits.webkit.org/259479@main">https://commits.webkit.org/259479@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6313175578d2f1c09adc058a152e2953b5f6c1ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104880 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13960 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114148 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174339 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15092 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4884 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97207 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113170 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110640 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11663 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94672 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39181 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93532 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80843 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94748 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7304 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27652 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92770 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5035 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7398 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4236 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30241 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47203 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101461 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9189 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25315 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3480 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

clopez commented on Tools/yocto/rpi/fix-nativesdk-image-defs.patch:
```diff
@@ -0,0 +1,406 @@
++           for using in the host with the script build-webkit and the image itself has \
++           the tools to run the tests on the board. But the image doesn't contain \
++           webkit (wpe) itself. The purpose of this image is for using on the \
++           continous integration systems of webkit.org for WPE tests."
```
Note: I also sent the diffs to meta-webkit as a PR here: https://github.com/Igalia/meta-webkit/pull/436
I will update this patch with the reduced delta after that PR in meta-webkit lands. But having the review there landed shouldn't block landing this

clopez replied:
PR merged into meta-webkit, I'm updating this patch with the reduced diff

aoikonomopoulos commented on Tools/Scripts/update-webkitgtk-libs:
```diff
@@ -24,6 +24,8 @@ use webkitdirs;
 if (defined $ENV{'WEBKIT_JHBUILD'} and $ENV{'WEBKIT_JHBUILD'}) {
     system("perl", "$scriptsDir/update-webkit-libs-jhbuild", "--gtk", @ARGV) == 0 or die $!;
+} elsif (defined $ENV{'WEBKIT_CROSS_TARGET'} or grep(/^--cross-target/, @ARGV)) {
+    system("perl", "$scriptsDir/cross-toolchain-helper", "--build-toolchain", @ARGV) == 0 or die $!;
```
Shouldn't this be `"python3"`?

clopez replied:
ups, yes.. :sweat_smile: 
good catch, i will fix it

clopez replied:
Funny enough this worked.
It seems that if you run `perl something` and that something has a shebang, then perl will execute it like if you ran it from a shell... TIL :teacher:

philn commented on Tools/yocto/README.md:
````diff
@@ -0,0 +1,258 @@
+3. Flash a SDCard with this image
+
+```
+user@workstation $ cat /path/to/core-image-webkit-dev-ci-tools-tests.wic.bz2 | bzip2 -dc | dd of=/dev/mmcblk0 status=progress bs=4k
````
This is not very fast/efficient. Maybe mention `bmaptool`, there's a nice blog post about it, https://blog.lazy-evaluation.net/posts/linux/bmaptool.html

clopez replied:
Good tip. I will add info about bmaptool to the README. Thanks

philn reviewed (commented):
What is the motivation for adding all this in the WebKit repo?

philn commented on Tools/yocto/README.md:
````diff
@@ -0,0 +1,268 @@
+user@workstation $ cat /path/to/core-image-webkit-dev-ci-tools-tests.wic.bz2 | bzip2 -dc | dd of=/dev/mmcblk0 status=progress bs=4k
+```
+
+4. Grow the rootfs partition to use all the available space on the SDcard
````
Is this step needed also when the rootfs is flashed with bmaptool? I'm not sure.

clopez replied:
It is needed. As far as I can see bmaptool does not grow the partition to the space available on the block.
Tested:

1. Clean the sdcard (just in case)
```
$ sudo dd if=/dev/zero of=/dev/mmcblk0 bs=1M count=20
20+0 records in
20+0 records out
20971520 bytes (21 MB, 20 MiB) copied, 0.34829 s, 60.2 MB/s

$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

2. Copy the image with bmaptool
```
$ sudo bmaptool copy webkit-dev-ci-tools.wic.bz2 /dev/mmcblk0
bmaptool: info: discovered bmap file 'webkit-dev-ci-tools.wic.bmap'
bmaptool: info: block map format version 2.0
bmaptool: info: 1069978 blocks of size 4096 (4.1 GiB), mapped 551797 blocks (2.1 GiB or 51.6%)
bmaptool: info: copying image 'webkit-dev-ci-tools.wic.bz2' to block device '/dev/mmcblk0' using bmap file 'webkit-dev-ci-tools.wic.bmap'
bmaptool: info: 100% copied
bmaptool: info: synchronizing '/dev/mmcblk0'
bmaptool: info: copying time: 39.7s, copying speed 54.2 MiB/sec
```
3. Show the partitions after bmaptool copy
```
$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x657f5e7b

Device         Boot  Start     End Sectors Size Id Type
/dev/mmcblk0p1 *      8192  112677  104486  51M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      114688 8559819 8445132   4G 83 Linux
```

As you see the last partition is only 4GBs (what is set on the image) instead of 60GB available on the sdcard.

If now I execute the other steps on the README then I get it to use all the space available.

```
$ sudo growpart /dev/mmcblk0 2
CHANGED: partition=2 start=114688 old: size=8445132 end=8559820 new: size=123482079,end=123596767

$ sudo resize2fs /dev/mmcblk0p2
resize2fs 1.44.5 (15-Dec-2018)
Resizing the filesystem on /dev/mmcblk0p2 to 15435259 (4k) blocks.
The filesystem on /dev/mmcblk0p2 is now 15435259 (4k) blocks long.


$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x657f5e7b

Device         Boot  Start       End   Sectors  Size Id Type
/dev/mmcblk0p1 *      8192    112677    104486   51M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      114688 123596766 123482079 58.9G 83 Linux
```

philn commented on Tools/yocto/targets.conf:
```diff
@@ -0,0 +1,77 @@
+patch_file_path = rpi/fix-nativesdk-image-defs.patch
+image_basename = webkit-dev-ci-tools
+image_extension = wic.bz2
+environment[BUILD_WEBKIT_ARGS] = --no-fatal-warnings --cmakeargs="-DENABLE_WPE_QT_API=OFF -DENABLE_THUNDER=OFF -DENABLE_DOCUMENTATION=OFF -DENABLE_INTROSPECTION=OFF -DWPE_COG_PLATFORMS=drm,headless,wayland"
```
meta-webkit includes a recipe for sparke-cdm, so maybe you can keep THUNDER on. You'd get some basic EME support at runtime, including clearkey layout tests coverage support.

clopez replied:
Great tip. I will do this. I have submitted a [PR to meta-webkit wit the changes required there](https://github.com/Igalia/meta-webkit/pull/444) and I will update this patch with the changes required there as well ASAP

clopez committed:
[WPE] Add support for generating a cross toolchain, images and cross building easily for a set of curated embedded targets with Yocto
https://bugs.webkit.org/show_bug.cgi?id=249604

Reviewed by Philippe Normand.

This adds a new script named 'cross-toolchain-helper' that
automates the process of building a cross-toolchain and an
image with Yocto.

The script also is able to automatically deploy the Yocto
workdir and set the target build configs from a set of
configuration files.

The main configuration file for this script is at the path
'Tools/yocto/targets.conf'. Inside this file several targets
can be defined. Each target defines a set of Yocto layers
and local.conf configurations. It can be also defined the
default WebKit parameters for each target.

The following WebKit tool scripts are also patched to be
aware of the cross-target builds:
  - Tools/Scripts/update-webkitwpe-libs
  - Tools/Scripts/update-webkitgtk-libs
  - Tools/Scripts/build-webkit
  - Tools/CISupport/built-product-archive

To enable the cross-target builds the flag '--cross-target=target_name'
should be passed. Alternatively the environment variable
'WEBKIT_CROSS_TARGET' can be used instead.

The list of targets defined at 'Tools/yocto/targets.conf'
is curated and is expected to build and work without issues.
In the near future a post-commit bot will be checking this.

The tool is designed to automatically wipe and recreate the
cross-toolchain and images when it detects the configuration
for any target has changed. However, this behaviour can be
overriden by optional parameters.

For more info (including examples on how to use this) check the file:
Tools/yocto/README.md

* Tools/CISupport/built-product-archive:
* Tools/Scripts/build-webkit:
* Tools/Scripts/cross-toolchain-helper: Added.
* Tools/Scripts/update-webkitgtk-libs:
* Tools/Scripts/update-webkitwpe-libs:
* Tools/Scripts/webkitdirs.pm:
(determineCrossTarget):
(determineConfigurationProductDir):
(determinePassedConfiguration):
(runInCrossTargetEnvironment):
(inCrossTargetEnvironment):
(getCrossTargetName):
(shouldBuildForCrossTarget):
(wrapperPrefixIfNeeded):
(shouldUseFlatpak):
(cmakeCachePath):
(cmakeFilesPath):
(shouldRemoveCMakeCache):
(cmakeGeneratedBuildfile):
(generateBuildSystemFromCMakeProject):
(productDirForCMake):
(buildCMakeGeneratedProject):
(cleanCMakeGeneratedProject):
* Tools/yocto/README.md: Added.
* Tools/yocto/rpi/bblayers.conf: Added.
* Tools/yocto/rpi/fix-nativesdk-image-defs.patch: Added.
* Tools/yocto/rpi/local-rpi3-32bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi3-32bits-userland.conf: Added.
* Tools/yocto/rpi/local-rpi3-64bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi4-32bits-mesa.conf: Added.
* Tools/yocto/rpi/local-rpi4-64bits-mesa.conf: Added.
* Tools/yocto/rpi/manifest.xml: Added.
* Tools/yocto/targets.conf: Added.

Canonical link: https://commits.webkit.org/259479@main
